src/libraries/System.Private.CoreLib/src/System/Globalization/Ordinal.cs

```diff
@@ -236,7 +238,46 @@ internal static int IndexOfOrdinalIgnoreCase(ReadOnlySpan<char> source, ReadOnly
                 return CompareInfo.NlsIndexOfOrdinalCore(source, value, ignoreCase: true, fromBeginning: true);
             }
 
-            return OrdinalCasing.IndexOf(source, value);
+            // if value starts with an ASCII char we can use a vectorized path
+            ref char valueRef = ref MemoryMarshal.GetReference(value);
+            char valueChar = valueRef;
+
+            if (!char.IsAscii(valueChar))
+            {
+                // Fallback to a more non-ASCII friendly version
+                return OrdinalCasing.IndexOf(source, value);
+            }
+
+            ref char searchSpace = ref MemoryMarshal.GetReference(source);
+            int valueLength = value.Length;
+            int searchSpaceLength = source.Length;
+
+            do
+            {
+                // if val is either [a..z] or [A..Z] - search for its lower and upper counter parts using IndexOfAny
+                // otherwise use just plain IndexOf
+                int candidatePos = (uint)((valueChar | 0x20) - 'a') <= 'z' - 'a' ?
+                    SpanHelpers.IndexOfAny(ref searchSpace, (char)(valueChar & ~0x20), (char)(valueChar | 0x20), searchSpaceLength) :
+                    SpanHelpers.IndexOf(ref searchSpace, valueChar, searchSpaceLength);
+
+                if (candidatePos == -1)
+                {
+                    // the whole input doesn't contain the first char
+                    goto NOT_FOUND;
+                }
+
+                // Do ASCII and non-ASCII friendly compare for the current candidate
+                if (EqualsIgnoreCase(ref searchSpace, ref valueRef, valueLength))
```

Comment:
Doesn't this comparison need to factor in candidatePos?

Reply:
oops! last minute change 🙂